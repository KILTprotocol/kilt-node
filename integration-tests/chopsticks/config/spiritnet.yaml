endpoint:
 - wss://kilt-rpc.dwellir.com
db: ./db/spiritnet.db.sqlite
port: ${env.SPIRITNET_PORT}
# wasm-override: ../../target/debug/wbuild/spiritnet-runtime/spiritnet_runtime.wasm
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000000"
