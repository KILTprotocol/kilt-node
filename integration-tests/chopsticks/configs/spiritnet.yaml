endpoint: wss://spiritnet.kilt.io
# mock-signature-host: true
db: out/spiritnet/db.sqlite
