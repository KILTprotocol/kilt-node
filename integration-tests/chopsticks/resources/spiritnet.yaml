endpoint: wss://spiritnet.kilt.io
mock-signature-host: true
db: ./db/spiritnet.db.sqlite
port: 8002

import-storage:
    technicalCommittee:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
    council:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
