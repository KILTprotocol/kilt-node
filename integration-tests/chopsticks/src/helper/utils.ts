import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
