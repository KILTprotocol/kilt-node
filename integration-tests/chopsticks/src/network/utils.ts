export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
