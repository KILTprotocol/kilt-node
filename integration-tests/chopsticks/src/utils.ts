import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
