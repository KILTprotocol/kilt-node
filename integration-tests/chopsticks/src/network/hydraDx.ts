import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceHDX, initialBalanceKILT, toNumber } from '../utils.js'

/// Options used to create the HydraDx context
export const options: SetupOption = {
	endpoint: process.env.HYDRADX_WS || ['wss://hydradx-rpc.dwellir.com', 'wss://rpc.hydradx.cloud'],
	db: './db/hydradx.db.sqlite',
	port: toNumber(process.env.HYDRADX_PORT) || 9001,
}

export const kiltTokenId = 28

/// Sets the [TechnicalCommittee] and [Council] governance to the given accounts
export function setGovernance(addr: string[]) {
	return {
		TechnicalCommittee: { Members: addr },
		Council: { Members: addr },
	}
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceHDX) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

/// Assigns KILT tokens to an accounts
export function assignKiltTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		Tokens: {
			Accounts: addr.map((address) => [[address, kiltTokenId], { free: balance }]),
		},
	}
}

/// HydraDX ParaId
export const paraId = 2034

/// Omnipool account
export const omnipoolAccount = '7L53bUTBbfuj14UpdCNPwmgzzHSsrsTWBHX5pys32mVWM3C1'

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
