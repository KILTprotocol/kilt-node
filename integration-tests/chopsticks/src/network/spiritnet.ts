import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceKILT, toNumber } from '../utils.js'

/// Options used to create the Spiritnet context
const options: SetupOption = {
	endpoint: process.env.SPIRITNET_WS || 'wss://kilt-rpc.dwellir.com',
	db: './db/spiritnet.db.sqlite',
	port: toNumber(process.env.SPIRITNET_PORT) || 9002,
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

/// Sets the [technicalCommittee] and [council] governance to the given accounts
export function setGovernance(addr: string[]) {
	return {
		technicalCommittee: { Members: addr },
		council: { Members: addr },
	}
}

export function XcmPalletSafeVersion3StorageEntry() {
	return {
		polkadotXcm: {
			safeXcmVersion: 3,
		},
	}
}

/// Spiritnet ParaId
export const paraId = 2086

/// The sovereign account of HydraDx in Spiritnet
export const hydraDxSovereignAccount = '4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt'

/// Returns the Spiritnet context for the given options
export async function getContext(): Promise<Config> {
	return setupContext(options)
}
