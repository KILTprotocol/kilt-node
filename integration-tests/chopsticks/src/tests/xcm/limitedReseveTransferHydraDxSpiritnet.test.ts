import { test } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import * as HydraDxConfig from '../../network/hydraDx.js'
import * as SpiritnetConfig from '../../network/spiritnet.js'
import { KILT, initialBalanceHDX, initialBalanceKILT, keysAlice } from '../../utils.js'
import { getFreeBalanceHydraDxKilt, getFreeBalanceSpiritnet, hydradxContext, spiritnetContext } from '../index.js'
import { checkBalance, createBlock, hexAddress, setStorage, checkBalanceInRange } from '../utils.js'

const aliceLocation = {
	V3: {
		parents: 1,
		interior: {
			X2: [
				{ Parachain: SpiritnetConfig.paraId },
				{
					AccountId32: {
						id: hexAddress(keysAlice.address),
					},
				},
			],
		},
	},
}

test('Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// assign initial balance to Alice. Alice also needs to have some HDX, otherwise the account gets dusted
	const config = {
		...HydraDxConfig.assignKiltTokensToAccounts([keysAlice.address], initialBalanceKILT),
		...HydraDxConfig.assignNativeTokensToAccounts([keysAlice.address], initialBalanceHDX),
	}

	// Set storage
	await setStorage(hydradxContext, config)

	const hydraDxSovereignAccountBalanceBeforeTransfer = await getFreeBalanceSpiritnet(
		SpiritnetConfig.hydraDxSovereignAccount
	)

	// check initial balance of alice
	await checkBalance(getFreeBalanceHydraDxKilt, keysAlice.address, expect, initialBalanceKILT)

	const signedTx = hydradxContext.api.tx.xTokens
		.transfer(HydraDxConfig.kiltTokenId, KILT, aliceLocation, 'Unlimited')
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	// Check sender state
	await createBlock(hydradxContext)

	// Check events sender
	checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender events xcm queue pallet')
	checkEvents(events, { section: 'currencies', method: 'Withdrawn' }).toMatchSnapshot('sender events currencies')
	checkEvents(events, 'xTokens').toMatchSnapshot('sender events currencies')

	// Check balance
	await checkBalance(getFreeBalanceHydraDxKilt, keysAlice.address, expect, initialBalanceKILT - KILT)

	// Check receiver state
	await createBlock(spiritnetContext)

	// check events receiver
	checkSystemEvents(spiritnetContext, 'xcmpQueue').toMatchSnapshot('receiver events xcmpQueue')
	checkSystemEvents(spiritnetContext, { section: 'balances', method: 'Withdraw' }).toMatchSnapshot(
		'receiver events Balances'
	)
	checkSystemEvents(spiritnetContext, { section: 'balances', method: 'Endowed' }).toMatchSnapshot(
		'receiver events Balances'
	)

	// Check balance receiver
	await checkBalance(
		getFreeBalanceSpiritnet,
		SpiritnetConfig.hydraDxSovereignAccount,
		expect,
		hydraDxSovereignAccountBalanceBeforeTransfer - KILT
	)

	// Alice receives a bit less since the tx fees has to be paid.
	await checkBalanceInRange(getFreeBalanceSpiritnet, keysAlice.address, expect, [
		BigInt('999999999971174'),
		BigInt('999999999976345'),
	])
}, 20_000)
