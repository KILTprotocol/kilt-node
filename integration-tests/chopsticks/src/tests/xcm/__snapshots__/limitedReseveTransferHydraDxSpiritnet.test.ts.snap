// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > receiver events Balances 1`] = `
[
  {
    "data": {
      "amount": 1000000000000000,
      "who": "4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt",
    },
    "method": "Withdraw",
    "section": "balances",
  },
]
`;

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > receiver events Balances 2`] = `
[
  {
    "data": {
      "account": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "freeBalance": "(rounded 1000000000000000)",
    },
    "method": "Endowed",
    "section": "balances",
  },
]
`;

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > receiver events xcmpQueue 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 800000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > sender events currencies 1`] = `
[
  {
    "data": {
      "amount": "(rounded 440000000000)",
      "currencyId": 0,
      "who": "7NL1GYCJu8cFSnWBLwET2X3fMdodw8T75zuxi59hA2bhdjQq",
    },
    "method": "Withdrawn",
    "section": "currencies",
  },
  {
    "data": {
      "amount": 1000000000000000,
      "currencyId": 28,
      "who": "7NL1GYCJu8cFSnWBLwET2X3fMdodw8T75zuxi59hA2bhdjQq",
    },
    "method": "Withdrawn",
    "section": "currencies",
  },
]
`;

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > sender events currencies 2`] = `
[
  {
    "data": {
      "assets": [
        {
          "fun": {
            "Fungible": 1000000000000000,
          },
          "id": {
            "Concrete": {
              "interior": {
                "X1": {
                  "Parachain": "(rounded 2100)",
                },
              },
              "parents": 1,
            },
          },
        },
      ],
      "dest": {
        "interior": {
          "X2": [
            {
              "Parachain": "(rounded 2100)",
            },
            {
              "AccountId32": {
                "id": "(hash)",
                "network": null,
              },
            },
          ],
        },
        "parents": 1,
      },
      "fee": {
        "fun": {
          "Fungible": 1000000000000000,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X1": {
                "Parachain": "(rounded 2100)",
              },
            },
            "parents": 1,
          },
        },
      },
      "sender": "7NL1GYCJu8cFSnWBLwET2X3fMdodw8T75zuxi59hA2bhdjQq",
    },
    "method": "TransferredMultiAssets",
    "section": "xTokens",
  },
]
`;

exports[`Limited Reserve Transfers from HydraDx Account Alice -> Spiritnet Account Alice > sender events xcm queue pallet 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
